nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,343 @@
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void SpeedFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      node->get_logger(), *(node->get_clock()), 2000,
+      "Filter mask was not received");
+    return;
+  }
+
+  double mask_robot_x, mask_robot_y;  // robot coordinates in mask frame
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Prepare transform from current layer frame (global_frame_) to mask frame
+    geometry_msgs::msg::TransformStamped transform;
+    tf2::Transform tf2_transform;
+    tf2_transform.setIdentity();  // initialize by identical transform
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "SpeedFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to (mask_robot_x, mask_robot_y) point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_robot_x = point.x();
+    mask_robot_y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_robot_x = pose.x;
+    mask_robot_y = pose.y;
+  }
+
+  // Converting (mask_robot_x, mask_robot_y) robot position
```

Comment:
For this block L253 - L271, can't this be replaced by a `worldToMap()`  call? Then a `getCost()` call with those coordinates rather than accessing `data` directly?

Reply:
`filter_mask_` in speed filter - is an OccupancyGrid map. It does not have any `worldToMap()` methods, we have for costmaps. Therefore, I've repeated the `worldToMap()`'s logic in here.

BTW this is nice small task for the future tickets - to write our own `OccupancyGrid` wrapper class in `nav2_utils` having all useful API from costmaps (like `worldToMap(), mapToWorld()`, `getCost()`, `getIndex()`, etc...) and use this class here instead. One hardening thing there - maps might be rotated - we need to count it when writing `worldToMap(), mapToWorld()` conversion routines. Some off-topic thoughts.

Reply:
Would the constructor you added in the original costmap filters PR not enable it? https://github.com/ros-planning/navigation2/blob/main/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp#L95

Reply:
Sure, we can do it. But I do not really think that it is good idea: to convert `OccupancyGrid` data `[0..100]` space -> to `Costmap2D` data `[0..255]` space -> back to percentage space `[0..100]`. This may lead to loss of accuracy. I think, that in speed limit filter direct usage of `OccupancyGrid` map data - is the best way to implement it.

Reply:
Ok, we don't have to do that, just a suggestion. I think a better course of action after looking over it again is to break out the `process` function into some more functions. Its unmaintainably long, so as part of that you can make a local mapToWorld for the occ grid message (among others)

Reply:
Made some process() logic into 2 helper functions `transformPose()` and `worldToMask()`: https://github.com/ros-planning/navigation2/pull/2074/files#diff-db4cc04c6df9717814a17beefe3d9013c83753db07ef4a6bfc47e4ce97240d31R79-R98
```